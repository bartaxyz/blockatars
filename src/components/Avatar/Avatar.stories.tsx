import { ComponentMeta, ComponentStory } from "@storybook/react";
import React from "react";

import { Avatar } from "./Avatar";

export default {
  title: "ReactComponentLibrary/Avatar",
  component: Avatar,
} as ComponentMeta<typeof Avatar>;

const Template: ComponentStory<typeof Avatar> = (args) => <Avatar {...args} />;

export const Base = Template.bind({});
Base.args = {
  size: 80,
};

export const Animated = Template.bind({});
Animated.args = {
  size: 80,
};
