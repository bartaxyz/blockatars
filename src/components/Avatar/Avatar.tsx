import React, { FC, useMemo } from "react";
import { lighten } from "polished";
import { useSeedColor } from "./hooks/useSeedColor";
import { useSeedRandom } from "./hooks/useSeedRandom";
import { Noise } from "../Noise/Noise";

export interface AvatarProps extends React.SVGProps<SVGSVGElement> {
  /**
   * A seed for the avatar (e.g. a wallet address, user id, etc.)
   */
  seed: string;

  /**
   * The display size of the avatar
   */
  size: number;

  /**
   * Disable automatic lower casing of the seed
   */
  disableLowerCase?: boolean;

  /**
   * Disable noise
   */
  disableNoise?: boolean;

  /**
   * Disable blur (for development & debugging purposes)
   */
  disableBlur?: boolean;
}

export const Avatar: FC<AvatarProps> = ({
  seed,
  size,
  disableLowerCase = false,
  disableNoise = false,
  disableBlur = false,
  ...props
}) => {
  const transformedSeed = disableLowerCase ? seed : seed.toLowerCase();
  const { shift } = useSeedRandom(transformedSeed);
  const { backgroundColor, color1, color2 } = useSeedColor(transformedSeed);

  const { shape1, shape2 } = useMemo(() => {
    const shape1 = shift();
    const shape2 = shift();

    const boundingBox = 200 / Math.sqrt(2);

    const shape1CenterX =
      (boundingBox / 2) * (1 + Math.cos(shape1 * 2 * Math.PI));
    const shape1CenterY =
      (boundingBox / 2) * (1 + Math.sin(shape1 * 2 * Math.PI));
    const shape1Radius = 80 + 20 * shift();

    const shape2CenterX =
      (boundingBox / 2) * (1 + Math.cos(shape2 * 2 * Math.PI));
    const shape2CenterY =
      (boundingBox / 2) * (1 + Math.sin(shape2 * 2 * Math.PI));
    const shape2Radius = 80 + 20 * shift();

    return {
      shape1: (
        <circle
          cx={shape1CenterX}
          cy={shape1CenterY}
          r={shape1Radius}
          fill={lighten(0.1, color1)}
        />
      ),
      shape2: (
        <circle
          cx={shape2CenterX}
          cy={shape2CenterY}
          r={shape2Radius}
          fill={lighten(0.1, color2)}
        />
      ),
    };
  }, [color1, color2, shift]);

  const scale = size / 200;

  return (
    <svg
      width={size}
      height={size}
      // viewBox="0 0 200 200"
      {...props}
      style={{
        /**
         * Keeping the noise image texture filter pixelated
         */
        imageRendering: "pixelated",
        ...props.style,
      }}
    >
      <defs>
        <filter id="blur">
          <feGaussianBlur in="SourceGraphic" stdDeviation="50" />
        </filter>
      </defs>

      <g transform={`scale(${scale})`}>
        <rect
          width={size / scale}
          height={size / scale}
          fill={lighten(0.25, backgroundColor)}
        />
      </g>

      <g transform={`scale(${scale})`}>
        <g filter={!disableBlur ? "url(#blur)" : undefined}>
          <rect width="100%" height="100%" fillOpacity="0" />
          {shape1}
          {shape2}
        </g>
      </g>

      {!disableNoise && <Noise />}
    </svg>
  );
};
