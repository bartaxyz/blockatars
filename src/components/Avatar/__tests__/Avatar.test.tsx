import React from "react";
import { render } from "@testing-library/react";

import { Avatar } from "../Avatar";

describe("Avatar", () => {
  test("renders the Avatar component", () => {
    const { asFragment } = render(<Avatar seed="0x1234567890" size={24} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
