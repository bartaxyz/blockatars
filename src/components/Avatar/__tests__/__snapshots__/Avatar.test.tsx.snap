// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders the Avatar component 1`] = `
<DocumentFragment>
  <svg
    height="24"
    viewBox="0 0 200 200"
    width="24"
  >
    <defs>
      <filter
        id=":r0:-blur"
      >
        <fegaussianblur
          in="SourceGraphic"
          stdDeviation="50"
        />
      </filter>
      <filter
        id=":r0:-noise"
      >
        <feturbulence
          baseFrequency="10"
          numOctaves="50"
          result="noise"
          type="fractalNoise"
        />
        <fecolormatrix
          result="saturate"
          type="saturate"
          values="1"
        />
        <fecomponenttransfer>
          <fefuncr
            intercept="0.5"
            slope="1"
            type="linear"
          />
          <fefuncg
            intercept="0.5"
            slope="1"
            type="linear"
          />
          <fefuncb
            intercept="0.5"
            slope="1"
            type="linear"
          />
          <fefunca
            type="identity"
          />
        </fecomponenttransfer>
        <fecomposite
          in2="SourceGraphic"
          operator="in"
        />
        <feblend
          in="rgba(0, 0, 0, 0.1)"
          in2="SourceGraphic"
          mode="multiply"
        />
      </filter>
    </defs>
    <g>
      <rect
        fill="#cf9782"
        height="100%"
        width="100%"
      />
      <g
        filter="url(#:r0:-blur)"
      >
        <rect
          fill-opacity="0"
          height="100%"
          width="100%"
        />
        <circle
          cx="138.55862081664222"
          cy="90.62691952235856"
          fill="#d0b77b"
          r="82.25128527730703"
        />
        <circle
          cx="130.26600180399788"
          cy="108.83105876646574"
          fill="#dabe7c"
          r="81.32641195319593"
        />
      </g>
    </g>
  </svg>
</DocumentFragment>
`;
